src/swarm-engine/Swarm/Game/CESK.hs

```diff
@@ -176,6 +163,11 @@ data Frame
     FRcd Env [(Var, Value)] Var [(Var, Maybe Term)]
   | -- | We are in the middle of evaluating a record field projection.
     FProj Var
+  | -- | We should suspend once we finish the current evaluation.
+    FSuspend Env
+  | -- | If an exception bubbles all the way up to this frame, then
+    --   switch to Suspended mode with this saved top-level context.
+    FRestoreEnv Env
```

Comment:
Hmm, good point... I think you're right, we can reuse `FSuspend` for both, which seems sensible.  I'll update it.

Reply:
Actually, after thinking about it more and trying to implement it, the answer is no, we need two different frames.  The difference is that an exception thrown inside a (genuine) `FSuspend` frame should ignore it and keep bubbling up until it reaches an enclosing fallback `FRestoreEnv` frame.  So we need to be able to distinguish the two.

Reply:
But while trying it out and thinking about it I did clean up + add a bunch more comments.

Reply:
@byorgey, could you give an example of how there could be more suspend/restore in the same stack?

In the original issue you even suggested not having a continuation in suspend. 🤔

Reply:
@xsebek normally, when the base starts executing a program entered at the REPL, we first put an `FRestoreEnv` frame on the stack, so that if the program crashes for some reason, we can at least restore the same environment we had before, so we do not lose any definitions that were made previously.  We also automatically insert an appropriate call to `suspend` somewhere at the end of the term.  For example if the term was `def x = blah end; return (foo bar)` we would rewrite it to `def x = blah end; suspend (foo bar)`.  When we get to executing the `suspend (foo bar)`, we do so by pushing an `FSuspend` frame onto the stack and then going on to evaluate `foo bar`.  So at that point there would be both an `FRestoreEnv` frame on the stack as well as an `FSuspend`.  If `foo bar` evaluates successfully then we will suspend with its result and discard the `FRestoreEnv`.  On the other hand if an uncaught exception is thrown during evaluation of `foo bar`, the exception will bubble up past the `FSuspend` and be caught by the `FRestoreEnv`.

Does that make sense?  I'm not sure whether I've answered your question.

Reply:
@byorgey I thought that suspend and restore are on the same side so to speak, but after using debug mode I guess they are more of inner/outer:

<img width="1125" alt="Screenshot 2024-06-18 at 19 25 22" src="https://github.com/swarm-game/swarm/assets/44544735/ab563bac-f5f1-4929-9204-57e8b2bd76ae">

I assume we could drop restore and check for end of stack of base, but maintaining a marker seems ok. 🤔

But there will be never more then one restore and one suspend and robots other than base do not get them. 👍

Reply:
> I assume we could drop restore and check for end of stack of base, but maintaining a marker seems ok.

No, the point of restore is not just to be a marker, it contains an environment (`Env`) with all the variables that were in scope + their values and types etc.  When a CESK machine returns (either normally via `Out` or with an exception via `Up`) it no longer contains an `Env`.  So if we did not save the original environment somewhere there would be no way to restore it.

Your next question might be, "why can't the CESK machine just remember an `Env` even in `Out` or `Up` states?"  but that would be incorrect --- it would mean that definitions made in local scopes would leak out into their parent scopes, i.e. #681.

Reply:
> I guess they are more of inner/outer

Right, `restore` will always be one of the very outermost things, whereas `suspend` gets inserted into the innermost scope at the end of all the binds and `let`/`def`.  This is so that programs entered at the REPL later will "see" the environment from that innermost scope, as if they were actually typed there.  In other words entering `a; b; c` at the REPL and then later `d; e; f` is the same as entering `a; b; c; d; e; f` (since bind is right-associative, note this is parenthesized as `a; (b; (c; (d; (e; f))))`.

Reply:
Thanks, @byorgey, for the long discussion.

At least in this short moment in time, it all makes sense to me. 😄